toqito/matrix_props/is_pseudo_unitary.py

```
@@ -0,0 +1,88 @@
+
+    .. math::
+        A^* J A = J,
+
+    where:
+        - :math:A^* is the conjugate transpose (Hermitian transpose) of :math:A,
+        - :math:J is a diagonal matrix with first p diagonal matrix equal to 1 and next q diagonal entries equal to -1
+
+    Examples
+    ==========
+
+    Consider the following matrix:
+
+    .. math::
+        A = \frac{1}{\sqrt{2}} \begin{pmatrix}
+            cosh(1) & sinh(1) \\
+            sinh(1) & cosh(1)
+        \end{pmatrix}
+
+    with the signature matrix:
+
+    .. math::
+        J = \begin{pmatrix}
+            1 & 0 \\
+            0 & -1
+        \end{pmatrix}
+
+    Our function confirms that :math:A is pseudo-unitary:
+
+    >>> import numpy as np
+    >>> from toqito.matrix_props import is_pseudo_unitary
+    >>> A = np.array([[np.cosh(1), np.sinh(1)], [np.sinh(1), np.cosh(1)]])
+    >>> is_pseudo_unitary(A, p=1, q=1)
+    np.True_
+
+    However, the following matrix :math:B
+
+    .. math::
+        B = \begin{pmatrix}
+            1 & 0 \\
+            1 & 1
+        \end{pmatrix}
+
+    is not pseudo-unitary with respect to the same signature matrix:
+
+    >>> B = np.array([[1, 0], [1, 1]])
+    >>> is_pseudo_unitary(B, p=1, q=1)
+    np.False_
+
+    References
+    ==========
+
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param mat: The matrix to check.
+    :param p: Number of positive entries in the signature matrix.
+    :param q: Number of negative entries in the signature matrix.
+    :param rtol: The relative tolerance parameter (default 1e-05).
+    :param atol: The absolute tolerance parameter (default 1e-08).
+    :raises ValueError: When p < 0 or q < 0.
+    :return: Return :code:True if the matrix is pseudo-unitary, and :code:False otherwise.
+
+    """
+    if p < 0 or q < 0:
+        raise ValueError("p and q must be non-negative")
+
+    if not is_square(mat):
+        return False
```

I wonder which option is better here:

- returning `False` for a check that is expected to be unsuccessful
- or raising a `ValueError`?

---

`ValueError` seems a better fit. But then should we also change the `is_unitary` function for consistency?

---

> ValueError seems a better fit

That's what I thought too. If the matrix is not square, then we already know it is not pseudounitary. 

> then should we also change the is_unitary function for consistency?

I think so. Pinging @vprusso for his thoughts.

---

But we already have this check for the `is_unitary` function:
https://github.com/vprusso/toqito/blob/master/toqito/matrix_props/is_unitary.py#L79

---

I'm not sure if `ValueError` is better. A non-square matrix shouldn't return a value error, as the value input is still a valid type (it's a numpy array). The matrix itself is just not unitary and should likely return `False` (instead of returning an error).

---

> But we already have this check for the is_unitary function

https://github.com/vprusso/toqito/blob/38ff1f52d2696e0f1580973f0ce337218824c984/toqito/matrix_props/is_unitary.py#L79-L80

Yes, I think @harshvardhan-pandey is referring to how the function returns false when the matrix is not square. That's the consistency being referred to. If we decide to a raise an error for this function then we should do the same for is_unitary.

---

> If we decide to a raise an error for this function then we should do the same for is_unitary.

Yes, but I don't think we should raise an error (if you refer to my follow-up comment above).

---

Fine by me then. I just think there's no need to use this function if we already know the matrix is not a square matrix.

---

Maybe I'm not following the conversation thread fully. 

>  I just think there's no need to use this function if we already know the matrix is not a square matrix.

Is "this function" `is_unitary`? Would we not want to check within that function whether it is square or not?

---

> Is "this function" is_unitary?

yes. 

> Would we not want to check within that function whether it is square or not?

Definitely want to keep this. 

All I am saying is it seems kinda obvious if a non-square matrix is not unitary i.e. we don't really need a function to check this. Your previous comment is fine by me though. Ok with returning `False` for both functions in this scenario.

---

@purva-thakre Then should the case where the matrix is square but it's not (p+q)x(p+q) also be flagged as False or should it be ValueError? Because even in then it is a valid input, it's just trivially non pseudo unitary.

---

Flag it as False to be consistent with Vincent's comment.